File: dsajava/src/main/java/com/leetcode/dsajava/my_daily_leetcode_challenge/backtrack_recursive/CoinChange_322.java
package com.leetcode.dsajava.my_daily_leetcode_challenge.backtrack_recursive;

import java.util.HashMap;
import java.util.Map;

public class CoinChange_322 {

    public static void main(String[] args) {
        int amount = 15;
        int[] coins = new int[]{2, 3, 7};
        System.out.println("res: " + coinChange(coins, amount));
    }

    public static int coinChange(int[] coins, int amount) {
        Map<Integer, Integer> memoiz = new HashMap<>();
        int res = backtrack(coins, amount, memoiz);
        if (res == Integer.MAX_VALUE / 2) return -1;
        return res;
    }

    public static int backtrack(int[] coins, int amount, Map<Integer, Integer> memoiz) {
        if (amount == 0) return 0;
        if (memoiz.get(amount) != null) return memoiz.get(amount);
        int min = Integer.MAX_VALUE / 2;
        for (int coin : coins) {
            if (amount - coin >= 0) {
                min = Math.min(min, 1 + backtrack(coins, amount - coin, memoiz));
            }
        }
        memoiz.put(amount, min);
        return min;
    }
}

File: dsa-kotlin/src/main/java/com/leetcode/dsakotin/my_daily_leetcode/dp/CoinChange_322.kt
package com.leetcode.dsakotin.my_daily_leetcode.dp

import kotlin.math.min


fun main() {

}

fun coinChange(coins: IntArray, amount: Int): Int {
    val MAX_VALUE = Int.MAX_VALUE/2
//    val dp = arrayOfNulls<Int>(amount+1).requireNoNulls() // for non-primitive
    val dp = IntArray(amount+1) // for primitive
//    for (i in 1 until amount+1) dp[i] = MAX_VALUE
//    (1..amount).forEach { index -> dp[index] = MAX_VALUE } // not necessary
    dp[0] = 0
    (1..amount).forEach { i ->
        var min = MAX_VALUE
        coins.forEach { coin ->
            if(i - coin >= 0) {
                min = min(min, 1 + dp[i-coin])
            }
        }
        dp[i] = min
    }
    return if(dp[amount] == MAX_VALUE) -1 else dp[amount]
}